### PR TITLE
Decouple Brave Ads resource, processor and model from text classification and purchase intent

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/13233

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [x] New files have MPL-2.0 license header.
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm visiting pages classifies the content. Confirm purchase intent is working as expected.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 1]
```diff
@@ -150,10 +161,17 @@ void AdsImpl::OnPageLoaded(
     last_visible_tab_url = last_visible_tab->url;
   }
 
-  purchase_intent_classifier_->MaybeExtractIntentSignal(url,
-      last_visible_tab_url);
+  if (!SameDomainOrHost(url, last_visible_tab_url)) {
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:164]: Not for now, but I'm on the fence again whether this piece of business logic should be responsibility of `AdsImpl` or the processor. Maybe a processor should observe tab updates to decide internally what action to take on subsequent visits for the same eTLD+1.

File: vendor/bat-native-ads/src/bat/ads/internal/search_engine/search_providers.cc [hunk 2]
```diff
@@ -5,6 +5,8 @@
 
 #include "bat/ads/internal/search_engine/search_providers.h"
 
+#include <iostream>
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/search_engine/search_providers.cc:8]: Nit: remove

**Review by moritzhaller — APPROVED** [review]
LGTM++

File: vendor/bat-native-ads/src/bat/ads/internal/search_engine/search_providers.cc (hunk 2 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/search_engine/search_providers.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc (hunk 1 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:164]: Agreed and we need to think of a better solution as right now a user can visit site a, change to site b and return to site a. Where site a would then be extracted again

Comment (moritzhaller) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:164]: moved to backlog since not for now

**Comment by bsclifton** [comment]
PR has been peer approved; CI looks good... iOS failure is due to https://github.com/brave/brave-browser/issues/13210 which has since been disabled I believe

Going to admin merge to unblock @moritzhaller 👍

**Pull request merged**